---
title: "Quality Control"
author: "Karol Nowicki-Osuch, based on Nils Eling code"
date: "`r Sys.Date()`"
output: 
html_document:
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: paper
    code_folding: hide
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '~/Dropbox/Postdoc/2019-12-29_BE2020/Results/Reports/1.4_Filtering.html') })
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
# export RSTUDIO_PANDOC=/usr/lib/rstudio-server/bin/pandoc
# Rscript -e "library(rmarkdown); render('1.4_Quality_control.Rmd')"
knitr::opts_chunk$set(echo = TRUE)
```

# Read in the data

We loop trough the folders containing the filtered matrix files, barcodes and genenames to read in the data in form of SingleCellExperiment (sce) objects.

```{r libraries, message=FALSE}
library(scran)
library(scater)
library(DropletUtils)
library(ggplot2)
library(Matrix)
library(cowplot)
library(openxlsx)
```

```{r additional_files}
MT_genes <- read.table("~/Dropbox/Postdoc/2019-12-29_BE2020/Additional_files/MT_genes.txt", sep = "\t", header = TRUE)
```

```{r data, eval=TRUE}
folders <- list.files("~/Dropbox/Postdoc/2019-12-29_BE2020/Data_v3/", full.names = TRUE)
folder.names <- list.files("~/Dropbox/Postdoc/2019-12-29_BE2020/Data_v3/", full.names = FALSE)

# Initialization of list to store sce objects
# Uncommand this when no other list exists
sce.list <- list()
# If other samples have already been analysed, load in list containing previous samples
# sce.list <- readRDS("~/Dropbox/Postdoc/2019-10-29_Gastric_IM/All_sce.rds")

# Read in the data
for(i in 1:length(folders)){
  if(folder.names[i] %in% names(sce.list)){next}
  else{
    cur_sce <- read10xCounts(paste0(folders[i],"/outs/filtered_feature_bc_matrix/"))
    colData(cur_sce)$Patient <- as.character(sapply(folder.names, 
                                         function(n){unlist(strsplit(n, "_"))[1]})[i])
    colData(cur_sce)$Tissue <- as.character(sapply(folder.names, 
                                        function(n){unlist(strsplit(n, "_"))[3]})[i])
    sce.list[[folder.names[i]]] <- cur_sce 
  }
}

# Save list
saveRDS(sce.list, "~/Dropbox/Postdoc/2019-12-29_BE2020/All_sce_unfiltered.rds")
```

# Plot QC features

Next we filter the cells based on several criteria. For this, we will plot these QC features first.

```{r QC_metrics}
#read in the file if running only this chunk
##sce.list <- readRDS("~/Dropbox/Postdoc/2019-10-29_Gastric_IM///All_sce_unfiltered.rds")
for(i in 1:length(sce.list)){
  cur_sce <- sce.list[[i]]
  if(!("total_features_by_counts" %in% names(colData(cur_sce)))){
    cur_sce <- suppressMessages(calculateQCMetrics(cur_sce))
    sce.list[[i]] <- cur_sce
  }
  
  # Library size
  print(ggplot(as.data.frame(colData(cur_sce))) + 
    geom_point(aes(1:ncol(cur_sce), log10(total_counts))) +
    xlab("Cells")+ggtitle(names(sce.list)[i]))
  
  # Number of genes detected
  print(ggplot(as.data.frame(colData(cur_sce))) + 
    geom_point(aes(1:ncol(cur_sce), total_features_by_counts)) +
    xlab("Cells")+ggtitle(names(sce.list)[i]))
  
  # Mitochondrial genes
  print(ggplot(data.frame(MT_genes = Matrix::colSums(counts(cur_sce)[rownames(cur_sce) %in%
                                                           MT_genes$Gene.stable.ID,])/
        Matrix::colSums(counts(cur_sce)))*100) +
    geom_point(aes(1:ncol(cur_sce), MT_genes)) + 
     ylab("% mitochondrial reads") + xlab("Cells")+ggtitle(names(sce.list)[i]))
  
  # Marker gene expression
  # VIM
  print(ggplot(data.frame(VIM = log10(counts(cur_sce)["ENSG00000026025",] + 1))) +
    geom_point(aes(1:ncol(cur_sce), VIM)) + xlab("Cells") +
    ylab("log10[VIM]")+ggtitle(names(sce.list)[i]))

  # PTPRC
  print(ggplot(data.frame(PTPRC = log10(counts(cur_sce)["ENSG00000081237",] + 1))) +
    geom_point(aes(1:ncol(cur_sce), PTPRC)) + xlab("Cells") +
    ylab("log10[PTPRC]")+ggtitle(names(sce.list)[i]))
}

# Create xlsx file with entries for QC thresholds
df <- data.frame(names = names(sce.list),
           lower_total_counts = rep(0, length(sce.list)),
           upper_total_counts = rep(0, length(sce.list)),
           lower_total_features = rep(0, length(sce.list)),
           upper_total_features = rep(0, length(sce.list)),
           lower_mito = rep(0, length(sce.list)),
           upper_mito = rep(0, length(sce.list)),
           VIM = rep(0, length(sce.list)),
           PTPRC = rep(0, length(sce.list)))

write.xlsx(df, "~/Dropbox/Postdoc/2019-12-29_BE2020/Additional_files/QC_metrics.xlsx")

saveRDS(sce.list, "~/Dropbox/Postdoc/2019-12-29_BE2020/All_sce_unfiltered.rds")
```

# Removal of low-quality cells

Here we remove cells based on the filtering thresholds defined by visualizing the QC features.

```{r filtering, eval = TRUE}
QC_thresholds <- read.xlsx("~/Dropbox/Postdoc/2019-12-29_BE2020/Additional_files/QC_metrics_edited.xlsx")
# sce.list <- readRDS("~/Dropbox/Postdoc/2019-10-29_Gastric_IM///All_sce_unfiltered.rds")

for(i in 1:length(sce.list)){
  cur_sce <- sce.list[[i]]
  
  # Total counts 
  cur_sce <- cur_sce[,log10(colData(cur_sce)$total_counts) >
                       QC_thresholds[i,"lower_total_counts"] &
                       log10(colData(cur_sce)$total_counts) <
                       QC_thresholds[i,"upper_total_counts"]]
  
  # Total features 
  cur_sce <- cur_sce[,colData(cur_sce)$total_features_by_counts >
                       QC_thresholds[i,"lower_total_features"] &
                       colData(cur_sce)$total_features_by_counts <
                       QC_thresholds[i,"upper_total_features"]]
  
  # Mitochondria 
  cur_sce <- cur_sce[,(Matrix::colSums(counts(cur_sce)[rownames(cur_sce) %in%
                MT_genes$Gene.stable.ID,])/Matrix::colSums(counts(cur_sce)))*100 >
                       QC_thresholds[i,"lower_mito"] &
                       (Matrix::colSums(counts(cur_sce)[rownames(cur_sce) %in%
                MT_genes$Gene.stable.ID,])/Matrix::colSums(counts(cur_sce)))*100 <
                       QC_thresholds[i,"upper_mito"]]
  
  # VIM
  if(!is.na(QC_thresholds[i,"VIM"])){
    cur_sce <- cur_sce[,counts(cur_sce)["ENSG00000026025",] <=
                         QC_thresholds[i,"VIM"]]
  }
  
  # PTPRC
  if(!is.na(QC_thresholds[i,"PTPRC"])){
    cur_sce <- cur_sce[,counts(cur_sce)["ENSG00000081237",] <=
                         QC_thresholds[i,"PTPRC"]]
  }
  sce.list[[i]] <- cur_sce
}

saveRDS(sce.list, "~/Dropbox/Postdoc/2019-12-29_BE2020/All_sce_filtered.rds")
```

# End Matter

To finish get session info:

```{r Endnote, include=TRUE, echo=TRUE, warning=FALSE, eval=TRUE, message=FALSE, tidy=TRUE}
sessionInfo()
```
