---
title: "Figure 1: overview on experimental approach and gland cells"
author: "Karol Nowicki-Osuch based on Nils Eling's code"
date: "`r Sys.Date()`"
output: 
html_document:
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: paper
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script visualizes the full dataset as well as highlighting the sub-populations in the gland cells.

# Read in the data

```{r libraries, message=FALSE}
library(scran)
library(scater)
library(DropletUtils)
library(openxlsx)
library(Rtsne)
library(pheatmap)
library(viridis)
library(umap)

source("../Analysis/Functions/auxiliary.R")

# Read in the normalized and batch-corrected reads
sce <- readRDS("~/Dropbox/Postdoc/2019-12-29_BE2020/All_corrected_sce_filtered.rds")
```

# Visualize gland cells

To visualize the gland cells, we perform batch correction across all patients and colour cells based on the clustering performed in the Tissue_correction script.

```{r}
cur_sce <- sce[,sce$Tissue == "SMG"]

# Split the SCE object into individual patients
sce.list <- split.sce(cur_sce, unique(cur_sce$Patient), colData.name = "Patient")

# Order sce object by size
sce.list <- sce.list[order(unlist(lapply(sce.list, ncol)), decreasing = TRUE)]

# Perform batch correction
corrected <- batch.correction(sce.list)

# Merge sce objects
cur_sce <- do.call("cbind", sce.list)

# Compute tSNE for gland cells
set.seed(111)
smg_tsne <- Rtsne(t(corrected), pca = FALSE)
umap <- umap(t(corrected))
# Create colour vecotr for glands
colour_vector <-  vector(length = length(unique(cur_sce$cell_type)))
names(colour_vector) <- unique(cur_sce$cell_type)
colour_vector["Immune"] <- "grey40"
colour_vector["Unknown.Doublets"] <- "white"
colour_vector["Mucous"] <- "saddlebrown"
colour_vector["Oncocytes"] <- "burlywood3"
colour_vector["Stromal"] <- "grey60"
# colour_vector["Nonepithelial"] <- "grey80"
colour_vector["Duct_Intercalating"] <- "burlywood4"
colour_vector["Squamous_Esophagus"] <- "black"
colour_vector["Myo-epithelial"] <- "brown"
# colour_vector["Dendritic"] <- "grey20"

p.smg <- ggplot(data.frame(tsne1 = smg_tsne$Y[,1],
                  tsne2 = smg_tsne$Y[,2],
                  cell_type = as.factor(cur_sce$cell_type))) +
  geom_point(aes(tsne1, tsne2), colour = "black", size = 2) + 
  geom_point(aes(tsne1, tsne2, colour = cell_type)) + 
  scale_color_manual(values = colour_vector) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.background = element_blank(),
        legend.key = element_blank()
  ) + 
  xlab("t-SNE 1")  + 
  ylab("t-SNE 2") + 
  guides(colour = guide_legend(override.aes = list(size=4), title = "Cell Type")
  )

p.smg

umap.smg <- ggplot(data.frame(UMAP1 = umap$layout[,1],
                  UMAP2 = umap$layout[,2],
                  cell_type = as.factor(cur_sce$cell_type))) +
  geom_point(aes(UMAP1, UMAP2), colour = "black", size = 2) + 
  geom_point(aes(UMAP1, UMAP2, colour = cell_type)) + 
  scale_color_manual(values = colour_vector) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.background = element_blank(),
        legend.key = element_blank()
  ) + 
  xlab("t-SNE 1")  + 
  ylab("t-SNE 2") + 
  guides(colour = guide_legend(override.aes = list(size=4), title = "Cell Type")
  )

umap.smg


ggplot(data.frame(tSNE1 = smg_tsne$Y[,1],
                  tSNE2 = smg_tsne$Y[,2],
                  cell_type = cur_sce$cell_type_secondary)) + 
  geom_point(aes(tSNE1, tSNE2), colour = "black", size = 2) +
  geom_point(aes(tSNE1, tSNE2, colour = cell_type)) +
  theme_minimal()

ggplot(data.frame(tSNE1 = smg_tsne$Y[,1],
                  tSNE2 = smg_tsne$Y[,2],
                  cell_type = as.factor(cur_sce$Tissue_cluster))) + 
  geom_point(aes(tSNE1, tSNE2), colour = "black", size = 2) +
  geom_point(aes(tSNE1, tSNE2, colour = cell_type)) +
  theme_minimal()

ggplot(data.frame(UMAP1 = umap$layout[,1],
                  UMAP2 = umap$layout[,2],
                  cell_type = cur_sce$cell_type_secondary)) + 
  geom_point(aes(UMAP1, UMAP2), colour = "black", size = 2) +
  geom_point(aes(UMAP1, UMAP2, colour = cell_type)) +
  theme_minimal()

ggplot(data.frame(UMAP1 = umap$layout[,1],
                  UMAP2 = umap$layout[,2],
                  cell_type = as.factor(cur_sce$Tissue_cluster))) + 
  geom_point(aes(UMAP1, UMAP2), colour = "black", size = 2) +
  geom_point(aes(UMAP1, UMAP2, colour = cell_type)) +
  theme_minimal()

ggsave("~/Dropbox/Postdoc/2019-12-29_BE2020/Figures/Figure_1/SMG_cells.pdf", 
       p.smg, 
       width = 8, height = 7, useDingbats = FALSE)
ggsave("~/Dropbox/Postdoc/2019-12-29_BE2020/Figures/Figure_1/SMG_cells_umap.pdf", 
       umap.smg, 
       width = 8, height = 7, useDingbats = FALSE)
```

# Find SMG marker genes

Here, we use edgeR to perform pairwise differential (DE) analysis between all SMG cell types.
The pairwise comparisons are then combined to find cell-type specific marker genes.

```{r}
# Remove non-expressed genes
cur_sce <- cur_sce[Matrix::rowSums(counts(cur_sce)) > 0,]

# Perform DE to find marker genes
DE.genes <- multi.DE(sce = cur_sce, 
                     conditions = cur_sce$cell_type,
                     covariate = cur_sce$Patient,
                     lfc = 0.5,
                     FDR = 0.1)

# We first normalize gene expression across all selected cells
cur_sce <- computeSumFactors(cur_sce, clusters=cur_sce$cell_type)
cur_sce <- logNormCounts(cur_sce, log = TRUE)

# Alternative DE
markers <- marker.detection(cur_sce, clusters = cur_sce$cell_type)

# compare to scran
match(DE.genes$Immune$Symbol, markers$Immune$GeneName)
match(DE.genes$Duct_Intercalating$Symbol, markers$Duct_Intercalating$GeneName)
match(DE.genes$Mucous$Symbol, markers$Mucous$GeneName)
match(DE.genes$Unknown.Doublets$Symbol, markers$Unknown.Doublets$GeneName)
match(DE.genes$Stromal$Symbol, markers$Stromal$GeneName)
# match(DE.genes$Nonepithelial$Symbol, markers$Nonepithelial$GeneName)
match(DE.genes$Squamous_Esophagus$Symbol, markers$Squamous_Esophagus$GeneName)
match(DE.genes$Oncocytes$Symbol, markers$Oncocytes$GeneName)
match(DE.genes$Myo_epithelial$Symbol, markers$`Myo-epithelial`$GeneName)
# match(DE.genes$Dendritic$Symbol, markers$Dendritic$GeneName)

# Visualize marker genes
cur_genes <- rownames(DE.genes$Mucous)
cur_counts <- logcounts(cur_sce)[cur_genes,order(cur_sce$cell_type)]
colnames(cur_counts) <- paste(cur_sce$Barcode, cur_sce$Patient, sep = "_")
pheatmap(as.matrix(cur_counts), scale = "row", 
         color = colorRampPalette(c("dark blue", "dark blue", "dark blue", "white", "dark red", "dark red", "dark red"))(100), cluster_rows = FALSE,
         cluster_cols = FALSE, 
         annotation_col = data.frame(row.names = colnames(cur_counts),
           cell_type = cur_sce$cell_type[order(cur_sce$cell_type)],
           patient = cur_sce$Patient[order(cur_sce$cell_type)]), 
         show_colnames = FALSE, show_rownames = FALSE, main = "Mucous markers")

# Visualize marker genes
cur_genes <- rownames(DE.genes$Oncocytes)
cur_counts <- logcounts(cur_sce)[cur_genes,order(cur_sce$cell_type)]
colnames(cur_counts) <- paste(cur_sce$Barcode, cur_sce$Patient, sep = "_")
pheatmap(as.matrix(cur_counts), scale = "row", 
         color = colorRampPalette(c("dark blue", "dark blue", "dark blue", "white", "dark red", "dark red", "dark red"))(100), cluster_rows = FALSE,
         cluster_cols = FALSE, 
         annotation_col = data.frame(row.names = colnames(cur_counts),
           cell_type = cur_sce$cell_type[order(cur_sce$cell_type)],
           patient = cur_sce$Patient[order(cur_sce$cell_type)]), 
         show_colnames = FALSE, show_rownames = FALSE, main = "Oncocytes markers")

# Visualize marker genes
cur_genes <- rownames(DE.genes$Duct_Intercalating)
cur_counts <- logcounts(cur_sce)[cur_genes,order(cur_sce$cell_type)]
colnames(cur_counts) <- paste(cur_sce$Barcode, cur_sce$Patient, sep = "_")
pheatmap(as.matrix(cur_counts), scale = "row", 
         color = colorRampPalette(c("dark blue", "dark blue", "dark blue", "white", "dark red", "dark red", "dark red"))(100), cluster_rows = FALSE,
         cluster_cols = FALSE, 
         annotation_col = data.frame(row.names = colnames(cur_counts),
           cell_type = cur_sce$cell_type[order(cur_sce$cell_type)],
           patient = cur_sce$Patient[order(cur_sce$cell_type)]), 
         show_colnames = FALSE, show_rownames = FALSE, main = "Duct markers")


# Visualize gene expression for control
ggplot(data.frame(tsne1 = smg_tsne$Y[,1],
                  tsne2 = smg_tsne$Y[,2],
                  gene = logcounts(cur_sce)[rowData(cur_sce)$Symbol == "KRT23",])) + 
  geom_point(aes(tsne1, tsne2, colour = gene)) +
  scale_colour_viridis()
ggplot(data.frame(UMAP1 = umap$layout[,1],
                  UMAP2 = umap$layout[,2],
                  gene = logcounts(cur_sce)[rowData(cur_sce)$Symbol == "KRT23",])) + 
  geom_point(aes(UMAP1, UMAP2, colour = gene)) +
  scale_colour_viridis()

# Save as table
write.xlsx(DE.genes, file = "/home/karolno/Dropbox/Postdoc/2019-12-29_BE2020/Figures/Supplemental_tables/Table_S2.xlsx")
```

# Heatmap visualizing SMG marker genes

Here, we display marker genes selected from the list of cluster-specific marker genes as generated by the Tissue_correction script.

```{r}
genes <- c("SLPI", "KRT7",
           "TFF3", "MUC5B", "AGR2", "KRT8",
           "MMP7", "SOX9", "CTNNB1", 
           "CDH1", "KRT23",
           "ACTA2", "KRT5", "TP63",
           "MKI67", "OLFM4")
  
# Remove somatic cells
cur_corrected <- corrected[,cur_sce$cell_type %in% 
                     c("Mucous", "Oncocytes", "Duct_Intercalating", "Myo-epithelial")]
sel_sce <- cur_sce[,cur_sce$cell_type %in% 
                     c("Mucous", "Oncocytes", "Duct_Intercalating", "Myo-epithelial")]
sel_sce <- logNormCounts(sel_sce)
for.heatmap <- logcounts(sel_sce)[match(genes, rowData(sel_sce)$Symbol),]
colnames(for.heatmap) <- paste(sel_sce$Barcode, sel_sce$Patient, sep = "_")

# Calculate euclidean distance using the factors after batch correction
euc.dist <- dist(t(cur_corrected), method = "euclidean")

# Colouring for patient
patient_vector <- c("dark red", "dark blue", "dark green")
names(patient_vector) <- c("Patient10", "Patient11", "Patient13")
pheatmap(for.heatmap, cluster_rows = FALSE, color = viridis(100), 
         labels_row = genes, show_colnames = FALSE, clustering_distance_cols = euc.dist,
         annotation_col = data.frame(row.names = colnames(for.heatmap),
                                     cell_type = sel_sce$cell_type,
                                     patient = sel_sce$Patient), 
         annotation_colors = list(cell_type = colour_vector,
                                  patient = patient_vector), 
         clustering_method = "ward.D2")
dev.off()
pdf(file = "/home/karolno/Dropbox/Postdoc/2019-12-29_BE2020/Figures/Figure_1/SMG_heatmap_lognorm.pdf", height = 7, width = 8, useDingbats = FALSE)
pheatmap(for.heatmap, cluster_rows = FALSE, color = viridis(100), 
         labels_row = genes, show_colnames = FALSE, clustering_distance_cols = euc.dist,
         annotation_col = data.frame(row.names = colnames(for.heatmap),
                                     cell_type = sel_sce$cell_type,
                                     patient = sel_sce$Patient), 
         annotation_colors = list(cell_type = colour_vector,
                                  patient = patient_vector), 
         clustering_method = "ward.D2")
dev.off()
```

# End Matter

To finish get session info:

```{r Endnote, include=TRUE, echo=TRUE, warning=FALSE, eval=TRUE, message=FALSE, tidy=TRUE}
sessionInfo()
```


