---
title: "Figure 1: overview on experimental approach and gland cells"
author: "Karol Nowicki-Osuch based on Nils Eling's code"
date: "`r Sys.Date()`"
output: 
html_document:
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: paper
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Here, I am visualising the data for the rebuttal. We are interested in the exrpesion of GATA6 and HNF4A

# Read in the data

```{r libraries, message=FALSE}
library(scran)
library(scater)
library(DropletUtils)
library(openxlsx)
library(Rtsne)
library(pheatmap)
library(cowplot)
library(RColorBrewer)
library(edgeR)
library(ape)
library(viridis)
library(umap)
library(reshape2)
source("../Analysis/Functions/auxiliary.R")
sce <- readRDS("~/Dropbox/Postdoc/2019-12-29_BE2020/All_corrected_sce_filtered.rds")

```

# Perform dimensionality reduction for NG and BE together

To find genes that are cell-type specific and differentially expressed between NG and BE, we focus the analysis only on these 2 tissues. I will be using immune cells for the analysis as anchors.

```{r}
# Select tissues 
cur_sce <- sce[,sce$include & sce$Tissue %in% c("NG", "BE", "NE")]

# Deselect cell types that do not match between tissues
# cur_sce <- cur_sce[,!(cur_sce$cell_type %in% c("Chief", "Endocrine", "Goblet", "Parietal", "Immune", "Stromal"))]
cur_sce <- cur_sce[,!(cur_sce$cell_type %in% c("Chief", "Endocrine", "Goblet", "Parietal"))]


# Remove genes that are not expressed
cur_sce <- cur_sce[Matrix::rowSums(counts(cur_sce)) > 0,]

# # Perform batch correction
# sce.list <- split.sce(cur_sce, unique(cur_sce$Sample), colData.name = "Sample")
# 
# # Order sce objects for batch correction
# sce.list <- sce.list[order(unlist(lapply(sce.list, ncol)), decreasing = TRUE)]
# sce.list <- sce.list[c(which(grepl("_NG_", names(sce.list))),
#                        which(grepl("_BE_", names(sce.list))))]
# 
# corrected <- batch.correction(sce.list)
# cur_sce <- do.call("cbind", sce.list)
n <- names(sort(table(cur_sce[["Sample"]]), decreasing = TRUE))
n <- n[c(which(grepl("_NE_", n)),
         which(grepl("_NG_", n)),
         which(grepl("_BE_", n)))]

# The samples are in NSCJ, NE, NG, SMG order  
corrected <- batch.correction.single(cur_sce, batches = "Sample", m.order = n)





#Identify the columns containing immune cells
nonepi<-!(colData(cur_sce)$cell_type %in% c("Immune", "Stromal"))
#All the subsequent images will not ahve immune and stromal cells on visualisations but they will be used for calculations

set.seed(1234)
```

# Visualize marker genes

```{r}
colour_vector <- vector(length = 13)
# names(colour_vector) <- unique(cur_sce$cell_type_secondary)
names(colour_vector) <-  c("Undifferentiated", "Undifferentiated_Dividing", "Foveolar_Intermediate", "Foveolar_differentiated","Columnar_Undifferentiated", "Columnar_Undifferentiated_Dividing", "Columnar_Intermediate", "Columnar_differentiated",
                          "Basal", "Suprabasal", "Suprabasal_Dividing", "Intermediate", "Superficial" )
colour_vector["Undifferentiated_Dividing"] <- colorRampPalette(c("white", "#3C5488FF"))(10)[3]
colour_vector["Foveolar_differentiated"] <- colorRampPalette(c("white", "#3C5488FF"))(10)[10]
colour_vector["Foveolar_Intermediate"] <- colorRampPalette(c("white", "#4DBBD5FF"))(10)[6]
colour_vector["Undifferentiated"] <- colorRampPalette(c("white", "#3C5488FF"))(10)[1]
# colour_vector["Stromal"] <- colorRampPalette(c("grey70", "#3C5488FF"))(10)[1]
# colour_vector["Immune"] <- colorRampPalette(c("grey90", "#3C5488FF"))(10)[1]
colour_vector["Columnar_differentiated"] <- colorRampPalette(c("white", "#3C5488FF"))(10)[10]
colour_vector["Columnar_Intermediate"] <- colorRampPalette(c("white", "#4DBBD5FF"))(10)[6]
colour_vector["Columnar_Undifferentiated"] <- colorRampPalette(c("white", "#3C5488FF"))(10)[1]
colour_vector["Columnar_Undifferentiated_Dividing"] <- colorRampPalette(c("white", "#3C5488FF"))(10)[3]
colour_vector["Basal"] <- colorRampPalette(c("white", "#3C5488FF"))(10)[1]
colour_vector["Suprabasal"] <- colorRampPalette(c("white", "#3C5488FF"))(10)[1]
colour_vector["Suprabasal_Dividing"] <- colorRampPalette(c("white", "#3C5488FF"))(10)[3]
colour_vector["Intermediate"] <- colorRampPalette(c("white", "#4DBBD5FF"))(10)[6]
colour_vector["Superficial"] <- colorRampPalette(c("white", "#3C5488FF"))(10)[10]


# genes <- c("MYC", "NCL", "CD44", "LEFTY1", 
#            "MGST2", "MGST3", "CHCHD10", "DNPH1", "FBP1",
#            "HNF4A", "KRT7",  "OCIAD2", "CLDN3")

genes <- c(
           "HNF4A", "GATA6")


# Normalization across all cells displayed
cur_sce <- computeSumFactors(cur_sce, clusters=paste(cur_sce$Tissue, cur_sce$cell_type_secondary))
cur_sce <- logNormCounts(cur_sce, log = TRUE)



# Visualize gene expression in form of boxplots
for.plot <- logcounts(cur_sce)[match(genes, rowData(cur_sce)$Symbol),nonepi]
rownames(for.plot) <- genes
colnames(for.plot) <- paste(cur_sce$Tissue[nonepi], cur_sce$cell_type_secondary[nonepi], cur_sce$Barcode[nonepi], cur_sce$Patient[nonepi], sep = "-")

library(reshape2)
for.plot.melt <- melt(as.matrix(for.plot))
for.plot.melt$Tissue <- sub("-.*$", "", for.plot.melt$Var2)
for.plot.melt$Tissue <- factor(for.plot.melt$Tissue, levels = c("NE", "NG", "BE"))
for.plot.melt$cell_type <- sapply(as.character(for.plot.melt$Var2), function(n){unlist(strsplit(n, "-"))[2]})
# for.plot.melt$cell_type <- factor(for.plot.melt$cell_type, levels = 
# c("Undifferentiated", "Dividing", "Foveolar_Intermediate", "Foveolar_differentiated", "Immune", "Stromal"))
for.plot.melt$cell_type <- factor(for.plot.melt$cell_type, levels = 
                                    c("Undifferentiated", "Undifferentiated_Dividing", "Foveolar_Intermediate", "Foveolar_differentiated","Columnar_Undifferentiated", "Columnar_Undifferentiated_Dividing", "Columnar_Intermediate", "Columnar_differentiated",
                                       "Basal", "Suprabasal", "Suprabasal_Dividing", "Intermediate", "Superficial"))

tissue.boxplot <- ggplot(for.plot.melt) + geom_boxplot(aes(x = Tissue, y = value, fill = cell_type)) +
  facet_wrap(~ Var1, nrow = 1) + scale_fill_manual(values = colour_vector)
# tissue.boxplot <- ggplot(for.plot.melt) + geom_violin(aes(x = Tissue, y = value, fill = cell_type)) +
# facet_wrap(~ Var1, nrow = length(genes)) + scale_fill_manual(values = colour_vector)
tissue.boxplot
ggsave("~/Dropbox/Postdoc/2019-12-29_BE2020/Figures/Supplementary_Note/Boxplot_HNF4A_GATA6.pdf", tissue.boxplot, width = 15, height = 5, useDingbats = FALSE)

```

# End Matter

To finish get session info:

```{r Endnote, include=TRUE, echo=TRUE, warning=FALSE, eval=TRUE, message=FALSE, tidy=TRUE}
sessionInfo()
```


